File: dart/lib/src/bindings/validate_record_v2_content.dart
import '../constants/addresses.dart';
import '../constants/records.dart';
import '../domain/get_domain_address.dart';
import '../errors/sns_errors.dart';
import '../instructions/instruction_types.dart';
import '../instructions/validate_roa_instructions.dart';

/// Parameters for validating record V2 content
class ValidateRecordV2ContentParams {
  const ValidateRecordV2ContentParams({
    required this.staleness,
    required this.domain,
    required this.record,
    required this.owner,
    required this.payer,
    required this.verifier,
  });

  /// Whether to perform staleness checking
  final bool staleness;

  /// The domain under which the record will be validated
  final String domain;

  /// The type of record to be validated
  final Record record;

  /// The address of the domain's owner
  final String owner;

  /// The address funding the operation
  final String payer;

  /// The address that will verify the record
  final String verifier;
}

/// Validates record V2 content with Solana signature validation.
///
/// This function mirrors js/src/bindings/validateRecordV2Content.ts
Future<TransactionInstruction> validateRecordV2Content(
  ValidateRecordV2ContentParams params,
) async {
  final domainResult = await getDomainAddress(
    GetDomainAddressParams(
      domain: '${params.record.value}.${params.domain}',
      record: RecordVersion.v2,
    ),
  );

  var parentAddress = domainResult.parentAddress;
  final isSub = domainResult.isSub;

  if (isSub) {
    final parentResult = await getDomainAddress(
      GetDomainAddressParams(domain: params.domain),
    );
    parentAddress = parentResult.domainAddress;
  }

  if (parentAddress == null) {
    throw InvalidParentError('Parent could not be found');
  }

  final instruction = ValidateRoaInstruction(staleness: params.staleness);

  return instruction.getInstruction(
    programAddress: recordsProgramAddress,
    systemProgram: systemProgramAddress,
    splNameServiceProgram: nameProgramAddress,
    feePayer: params.payer,
    record: domainResult.domainAddress,
    domain: parentAddress,
    domainOwner: params.owner,
    centralState: centralStateDomainRecords,
    verifier: params.verifier,
  );
}

File: js/src/bindings/validateRecordV2Content.ts
import { SNS_RECORDS_ID, validateSolanaSignature } from "@bonfida/sns-records";
import { PublicKey } from "@solana/web3.js";

import { NAME_PROGRAM_ID } from "../constants";
import { InvalidParentError } from "../error";
import { Record, RecordVersion } from "../types/record";
import { getDomainKeySync } from "../utils/getDomainKeySync";
import { parseSupportedTld, SNS_TLD } from "../utils/tld";

/**
 * Builds the instruction to validate a V2 record's content using a Solana
 * signature, proving that the record was set by the domain owner or an
 * authorised verifier.
 *
 * @param staleness When `true`, validates the staleness signature (proving the
 *   record content matches the current owner); when `false`, validates the
 *   Right of Association signature instead.
 * @param domain The full domain name including TLD (e.g. `"mydomain.sns"`)
 * @param record The record type to validate
 * @param owner The owner of the domain
 * @param payer The fee payer of the transaction
 * @param verifier The account whose signature is being verified
 * @returns A {@link TransactionInstruction} that validates the record content
 * @throws {InvalidParentError} When the parent domain account cannot be resolved
 */
export const validateRecordV2Content = (
  staleness: boolean,
  domain: string,
  record: Record,
  owner: PublicKey,
  payer: PublicKey,
  verifier: PublicKey,
) => {
  // Only allows .sns domains
  parseSupportedTld(domain, [SNS_TLD]);

  let { pubkey, parent, isSub } = getDomainKeySync(
    `${record}.${domain}`,
    RecordVersion.V2,
  );

  if (isSub) {
    parent = getDomainKeySync(domain).pubkey;
  }

  if (!parent) {
    throw new InvalidParentError("Parent could not be found");
  }

  const ix = validateSolanaSignature(
    payer,
    pubkey,
    parent,
    owner,
    verifier,
    NAME_PROGRAM_ID,
    staleness,
    SNS_RECORDS_ID,
  );
  return ix;
};
